File: src/shared/ai/conversation/origins.js
// Original source IDs and current-request wire IDs are different namespaces.
const CONVERSATION_ID=/^I([1-9][0-9]{0,6})_P([0-9]{1,6})$/;
const LEGACY_WIRE_ID=/^P[0-9]{1,6}$/;
const originalId = value => typeof value === 'string' && Array.from(value).length > 0 &&
  Array.from(value).length <= 160 && value === value.trim() && !/[\x00-\x1f\x7f-\x9f\ud800-\udfff]/u.test(value);
function invalid(field, reason) {
  throw Object.assign(new Error('Conversation source mapping is invalid'), {
    code: 'ai_conversation_origin_invalid', stage: 'conversation_mapping', origin: 'client',
    category: 'input', retryable: false, providerAttempts: 0, generationAttempts: 0,
    requestDispatched: false, validation: {field, reason},
  });
}
export function checkedOrigins(value, sourceIds = null) {
  if (value == null) return [];
  if (!Array.isArray(value) || value.length > 128) invalid('conversation.origins','invalid_page_list');
  const seen = new Set(), pages = new Set();
  const rows = value.map((item,index) => {
    const field = `conversation.origins.${index}`;
    if (!item || typeof item !== 'object' || Array.isArray(item)) invalid(field,'invalid_page_origin');
    const ids = item.unitIds, originals = item.originalIds;
    if (!Array.isArray(ids) || !ids.length || ids.length > 2000 || !Array.isArray(originals) || ids.length !== originals.length)
      invalid(field,'incomplete_mapping');
    for (const [i,id] of ids.entries()) {
      if (typeof id !== 'string' || !(CONVERSATION_ID.test(id)||LEGACY_WIRE_ID.test(id))) invalid(`${field}.unitIds.${i}`,'invalid_wire_id');
      const match=CONVERSATION_ID.exec(id);
      if(match&&Number.isInteger(item.pageOrder)&&Number(match[1])!==item.pageOrder) invalid(`${field}.unitIds.${i}`,'image_id_mismatch');
      if (seen.has(id)) invalid(`${field}.unitIds.${i}`,'duplicate_wire_id');
      seen.add(id);
    }
    const local = new Set();
    for (const [i,id] of originals.entries()) {
      if (!originalId(id)) invalid(`${field}.originalIds.${i}`,'invalid_source_id');
      if (local.has(id)) invalid(`${field}.originalIds.${i}`,'duplicate_source_id');
      local.add(id);
    }
    if (seen.size > 2000) invalid('conversation.origins','too_many_wire_ids');
    if (!originalId(item.pageId)) invalid(`${field}.pageId`,'invalid_page_id');
    if (pages.has(item.pageId)) invalid(`${field}.pageId`,'duplicate_page_id');
    pages.add(item.pageId);
    const fingerprint = Object.prototype.hasOwnProperty.call(item, 'sourceFingerprint') ? item.sourceFingerprint : '';
    if (typeof fingerprint !== 'string' || fingerprint && !/^[a-f0-9]{64}$/.test(fingerprint)) invalid(`${field}.sourceFingerprint`,'invalid_fingerprint');
    const row = {pageId:item.pageId,unitIds:[...ids],originalIds:[...originals],sourceFingerprint:fingerprint};
    for(const key of ['pageIndex','pageOrder']) if(Number.isInteger(item[key]) && item[key]>=0 && item[key]<10000000) row[key]=item[key];
    return row;
  });
  if (rows.length && sourceIds && (seen.size !== sourceIds.length || rows.flatMap(p=>p.unitIds).some((id,i)=>id!==sourceIds[i])))
    invalid('conversation.origins','source_order_mismatch');
  return rows;
}

// Page ownership only. Old assistant strings must stay byte-identical.
const key=p=>p.pageIndex!=null?`index:${p.pageIndex}`:`id:${p.pageId}`;
export function branchHistory(history,origins,orderPolicy="request_arrival") {
  for(let i=0;i<history.length;i++)for(const old of history[i].pages||[])for(const fresh of origins){
    if(key(old)===key(fresh)){
      const changed=old.sourceFingerprint&&fresh.sourceFingerprint&&old.sourceFingerprint!==fresh.sourceFingerprint;
      if(changed||old.originalIds?.some(id=>fresh.originalIds?.includes(id)))return {turns:history.slice(0,i),reason:changed?'source_changed':'source_replayed'};
    }else if(orderPolicy==='document_enqueue'&&old.pageIndex!=null&&fresh.pageIndex!=null&&old.pageIndex>fresh.pageIndex)return {turns:history.slice(0,i),reason:'source_order_rewound'};
  }
  return {turns:history,reason:'none'};
}
export function pageBoundaries(origins,locale){
  // I<image>_P<unit> carries page ownership in the record ID itself. Normal
  // Conversation continuations therefore need no repeated page/speaker prose.
  if(origins.every(p=>(p.unitIds||[]).every(id=>CONVERSATION_ID.test(id)))) return '';
  if(origins.length<2)return '';
  const title={th:'ขอบเขตภาพ — ID แต่ละกลุ่มเป็นคนละภาพ ไม่ได้ระบุผู้พูด',ja:'ページ境界 — IDの組は別の画像です。話者を表しません。'}[locale]||'Page boundaries — each ID group belongs to a separate image, not a speaker';
  return title+'\n'+origins.map((p,i)=>`${i+1}: ${p.unitIds.join(', ')}`).join('\n')+'\n\n';
}

File: api/backend/ai/translation_paths/origins.py
"""Page/unit ownership and automatic replay branching; never edits old answers."""
from __future__ import annotations
import hashlib
import json
import re


MAX_ORIGINAL_ID_CHARS = 160
CONVERSATION_ID_RE = re.compile(r"I([1-9][0-9]{0,6})_P([0-9]{1,6})")
LEGACY_WIRE_ID_RE = re.compile(r"P[0-9]{1,6}")


class OriginValidationError(ValueError):
    """Typed mapping error; diagnostics describe structure, never source values."""
    code = "ai_conversation_origin_invalid"
    stage = "conversation_mapping"
    requestDispatched = False
    providerAttempts = 0
    generationAttempts = 0

    def __init__(self, field, reason):
        self.validation = {"field": field, "reason": reason}
        super().__init__("Conversation source mapping is invalid")


def _original_id(value):
    # Original IDs belong to the OCR/document, not to the model's Pn protocol.
    # Keep them opaque and case-sensitive (g0, p0, UUID, etc.), without coercion.
    return (isinstance(value, str) and 0 < len(value) <= MAX_ORIGINAL_ID_CHARS
            and value == value.strip()
            and not re.search(r"[\x00-\x1f\x7f-\x9f\ud800-\udfff]", value))


def checked_origins(value):
    if value is None:
        return []
    if not isinstance(value, list) or len(value) > 128:
        raise OriginValidationError("conversation.origins", "invalid_page_list")
    result, seen, pages = [], set(), set()
    for index, item in enumerate(value):
        field = f"conversation.origins.{index}"
        if not isinstance(item, dict):
            raise OriginValidationError(field, "invalid_page_origin")
        ids, originals = item.get("unitIds"), item.get("originalIds")
        if (not isinstance(ids, list) or not ids or len(ids) > 2000
                or not isinstance(originals, list) or len(ids) != len(originals)):
            raise OriginValidationError(field, "incomplete_mapping")
        local = set()
        for position, uid in enumerate(ids):
            if not isinstance(uid, str) or not (CONVERSATION_ID_RE.fullmatch(uid) or LEGACY_WIRE_ID_RE.fullmatch(uid)):
                raise OriginValidationError(f"{field}.unitIds.{position}", "invalid_wire_id")
            match = CONVERSATION_ID_RE.fullmatch(uid)
            if match and isinstance(item.get("pageOrder"), int) and int(match.group(1)) != item["pageOrder"]:
                raise OriginValidationError(f"{field}.unitIds.{position}", "image_id_mismatch")
            if uid in seen or uid in local:
                raise OriginValidationError(f"{field}.unitIds.{position}", "duplicate_wire_id")
            local.add(uid)
        original_seen = set()
        for position, uid in enumerate(originals):
            if not _original_id(uid):
                raise OriginValidationError(f"{field}.originalIds.{position}", "invalid_source_id")
            if uid in original_seen:
                raise OriginValidationError(f"{field}.originalIds.{position}", "duplicate_source_id")
            original_seen.add(uid)
        seen.update(local)
        if len(seen) > 2000:
            raise OriginValidationError("conversation.origins", "too_many_wire_ids")
        page = item.get("pageId")
        if not _original_id(page):
            raise OriginValidationError(f"{field}.pageId", "invalid_page_id")
        if page in pages:
            raise OriginValidationError(f"{field}.pageId", "duplicate_page_id")
        pages.add(page)
        row = {"pageId": page, "unitIds": list(ids), "originalIds": list(originals)}
        for key in ("pageIndex", "pageOrder"):
            v = item.get(key)
            if isinstance(v, int) and not isinstance(v, bool) and 0 <= v < 10_000_000:
                row[key] = v
        fingerprint = item.get("sourceFingerprint", "")
        if not isinstance(fingerprint, str) or (fingerprint and not re.fullmatch(r"[a-f0-9]{64}", fingerprint)):
            raise OriginValidationError(f"{field}.sourceFingerprint", "invalid_fingerprint")
        row["sourceFingerprint"] = fingerprint
        result.append(row)
    return result


def validate_source_order(rows, source_ids):
    """At the API batch ingress IDs already name the current Pn wire sequence."""
    if rows and [uid for row in rows for uid in row["unitIds"]] != list(source_ids):
        raise OriginValidationError("conversation.origins", "source_order_mismatch")
    return rows


def current_origins(description, source_texts):
    rows = checked_origins(description.get("origins"))
    if rows:
        return rows
    if not description.get("pageId"):
        return []
    return [{**{k:description[k] for k in ("pageId", "pageIndex", "pageOrder") if k in description},
        "unitIds":[f"P{i}" for i in range(len(source_texts))],
        "originalIds":[f"P{i}" for i in range(len(source_texts))],
        "sourceFingerprint":hashlib.sha256(json.dumps(source_texts,ensure_ascii=False).encode()).hexdigest()}]


def branch_history(history, origins, order_policy="request_arrival"):
    def page_key(p):
        return ("index",p["pageIndex"]) if p.get("pageIndex") is not None else ("id",p.get("pageId"))
    for index, turn in enumerate(history):
        for old in turn.get("pages", []):
            for new in origins:
                if page_key(old) == page_key(new):
                    changed = old.get("sourceFingerprint") and new.get("sourceFingerprint") and old["sourceFingerprint"] != new["sourceFingerprint"]
                    if changed or set(old.get("originalIds", [])) & set(new.get("originalIds", [])):
                        return history[:index], "source_changed" if changed else "source_replayed"
                elif order_policy == "document_enqueue" and old.get("pageIndex") is not None and new.get("pageIndex") is not None and old["pageIndex"] > new["pageIndex"]:
                    return history[:index], "source_order_rewound"
    return history, "none"


def boundaries(origins, locale):
    # Stable image/unit IDs make repeated page-boundary prose unnecessary.
    if origins and all(CONVERSATION_ID_RE.fullmatch(uid) for row in origins for uid in row.get("unitIds", [])):
        return ""
    if len(origins) < 2:
        return ""
    title = {"th":"ขอบเขตภาพ — ID แต่ละกลุ่มเป็นคนละภาพ ไม่ได้ระบุผู้พูด", "ja":"ページ境界 — IDの組は別の画像です。話者を表しません。"}.get(locale,
        "Page boundaries — each ID group belongs to a separate image, not a speaker")
    return title + "\n" + "\n".join(f"{i+1}: " + ", ".join(row["unitIds"]) for i,row in enumerate(origins)) + "\n\n"
